xmtp_mls/src/identity/xmtp_id/identity.rs

```diff
@@ -71,112 +71,129 @@ impl Identity {
+                builder = builder.create_inbox(member_identifier.clone(), nonce);
 
-        let signature_keys = SignatureKeyPair::new(CIPHERSUITE.signature_algorithm())?;
-        let installation_public_key = signature_keys.public();
-        let member_identifier: MemberIdentifier = address.into();
-        let builder = SignatureRequestBuilder::new(inbox_id.clone());
-        let mut signature_request = builder
-            .create_inbox(member_identifier.clone(), nonce)
-            .add_association(installation_public_key.to_vec().into(), member_identifier)
-            .build();
+                let mut signature_request = builder
+                    .add_association(installation_public_key.to_vec().into(), member_identifier)
+                    .build();
 
-        // We can pre-sign the request with an installation key signature, since we have access to the key
-        signature_request
-            .add_signature(Box::new(
-                sign_with_installation_key(
-                    signature_request.signature_text(),
-                    sized_installation_key(signature_keys.private())?,
-                )
-                .await?,
-            ))
-            .await?;
+                // We can pre-sign the request with an installation key signature, since we have access to the key
+                signature_request
+                    .add_signature(Box::new(
+                        sign_with_installation_key(
+                            signature_request.signature_text(),
+                            sized_installation_key(signature_keys.private())?,
+                        )
+                        .await?,
+                    ))
+                    .await?;
 
-        signature_request
-            .add_signature(Box::new(
-                sign_with_legacy_key(
-                    signature_request.signature_text(),
-                    legacy_signed_private_key,
-                )
-                .await?,
-            ))
-            .await?;
+                let identity = Self {
+                    inbox_id: inbox_id.clone(),
+                    installation_keys: signature_keys,
+                    credential: create_credential(inbox_id.clone())?,
+                    signature_request: Some(signature_request),
+                };
 
-        let identity_update = signature_request.build_identity_update()?;
+                Ok(identity)
+            }
+        } else {
+            // If an inbox is associated, we just need to associate the installation key
+            // Only wallet is allowed to sign the installation key, legacy key is not
+            let nonce = rand::random::<u64>();
+            let inbox_id = generate_inbox_id(&address, &nonce);
+            let builder = SignatureRequestBuilder::new(inbox_id.clone());
+            let mut signature_request = builder
+                .add_association(installation_public_key.to_vec().into(), member_identifier)
+                .build();
```

Comment:
I'm not sure I understand this bit of code.

If there is no existing `inbox_id` and we are randomly generating the nonce, we would need to create the inbox with the random nonce. Can't just do an `add_association` to the inbox that hasn't been created yet.

Reply:
this `else {}` part is for existing `inbox_id`.

I also find this code structure confusing, & tedious, will make this DRYer.